wled00/espnow_broadcast.cpp

```diff
@@ -0,0 +1,249 @@
+#ifdef NODE_DEBUGGING
+                                Serial.println("atomic state out of sync");
+#endif
+                            }
+                        } else {
+#ifdef NODE_DEBUGGING
+                            Serial.println("esp_now_add_peer failed");
+#endif
+                        }
+                    } else {
+#ifdef NODE_DEBUGGING
+                        Serial.println("esp_now_register_recv_cb failed");
+#endif
+                    }
+                } else {
+#ifdef NODE_DEBUGGING
+                    Serial.println("esp_now_init_init failed");
+#endif
+                }
+            } else {
+#ifdef NODE_DEBUGGING
+                Serial.println("esp_wifi_start failed");
+#endif
+            }
+        } else {
+#ifdef NODE_DEBUGGING
+            Serial.println("WiFi.status not disconnected");
+#endif
+        }
+    } else {
+#ifdef NODE_DEBUGGING
+        Serial.println("WiFi.mode failed");
+#endif
+    }
+    Serial.println("restarting ESPNow");
+    setupWiFi();
+}
+
+
+void ESPNOWBroadcast::onWiFiEvent(void* arg, esp_event_base_t event_base, int32_t event_id, void* event_data) {
+    if ( event_base == WIFI_EVENT ) {
+#ifdef NODE_DEBUGGING
+        Serial.printf("WiFiEvent( %S )\n", WiFi.eventName(event) );
+#endif
+        switch (event_id) {
+            case WIFI_EVENT_STA_START: {
+                ESPNOWBroadcast::STATE stopped {ESPNOWBroadcast::STOPPED};
+                espnowBroadcast._state.compare_exchange_strong(stopped, ESPNOWBroadcast::STARTING);            
+                break;
+            }
+
+            case WIFI_EVENT_STA_STOP:
+            case WIFI_EVENT_AP_START: {
+                ESPNOWBroadcast::STATE started {ESPNOWBroadcast::STARTED};
+                ESPNOWBroadcast::STATE starting {ESPNOWBroadcast::STARTING};
+                if (espnowBroadcast._state.compare_exchange_strong(started, ESPNOWBroadcast::STOPPED) ||
+                    espnowBroadcast._state.compare_exchange_strong(starting, ESPNOWBroadcast::STOPPED)) {
+#ifdef NODE_DEBUGGING                            
+                    Serial.println("WiFi connected: stop broadcasting");
+#endif                        
+                    esp_err_t err = esp_now_deinit();
+                }
+                break;
+            }
+        }
+    }
+}
+
+bool ESPNOWBroadcast::registerCallback( receive_callback_t callback ) {
+    // last element is always null
+    size_t ndx;    
+    for (ndx = 0; ndx < _rxCallbackSize-1; ndx++) {
+        if (nullptr == _rxCallback[ndx]) {
+            _rxCallback[ndx] = callback;
+            break;
+        }
+    }
+    return ndx < _rxCallbackSize;
+}
+
+bool ESPNOWBroadcast::removeCallback( receive_callback_t callback ) {
+    size_t ndx;
+    for (ndx = 0; ndx < _rxCallbackSize-1; ndx++) {
+        if (_rxCallback[ndx] == callback ) {
+            break;
+        }
+    }
+
+    for (; ndx < _rxCallbackSize-1; ndx++) { 
+        _rxCallback[ndx] == _rxCallback[ndx+1];

```

Comment:
this is probably not true ;)

Reply:
correct. caught that as I'm reworking the ringbuf